### ISSUE TITLE
Cannot identify user (when blocking main thread?)

### ISSUE THREAD

**Issue by mateuszkwiecinski** [launchdarkly/android-client-sdk#77]
**Describe the bug**
I've migrated my app from 2.7.0 to 2.8.3 and I started observing wrong behaviour when calling `LDClientInterface#indentify` - the app started hanging because toggle call does not complete anymore. 
I have my own RxJava2 wrapper for LDClient library which looks follows:
```
Completable.create {
                    val future = client.identify(newUser)

                    try {
                        future.get()
                        it.onComplete()
                    } catch (exception: InterruptedException) {
                        if (!it.isDisposed) {
                            it.onError(exception)
                        }
                    }
                    it.setCancellable { future.cancel(true) }
                }.subscribeOn(Schedulers.io())
```
 Starting from LaunchDarkly client version 2.8.0 the ` future.get()` never completes.

**To reproduce**
As for now I found out that to reproduce wrong behaviour it is enoug to call `blockingAwait` on returned Completable from above snippet. (I tried to call it directly from applicataion class, but it doesn't work when called from Fragment class too)
I'll try to provide sample project with smallest possible repro, it should be reproducible withouth rxjava wrapper too.

**Expected behavior**
I'd like the Future returned by `indetify` to complete at some point. 

**Logs**
I tried to register ConnectionInformation observer but hasn't been ever called when the problem occurs. 

**SDK version**
I've tried 2.7.0 (works), 2.8.0, 2.8.1, 2.8.3 (it doesn't work)

**Language version, developer tools**
AGP 3.4.1
Kotlin 1.3.31

**OS/platform**
Tested on Android api 26, 28.

**Additional context**
- My main app does not call `blockingAwait()` or ever blocks the thread. Repro steps I provided I made only in smaller app.
- Additional interesting thing is that LD version 2.8.X does not work with my main app only when I clear app data completely (`android:allowBackup` is set to false). Otherwise, it works properly sometimes for few people. So by reporting this bug I'm hoping those are somehow connected because of some race condition. 
- And as a confirmation, my general LD setup is correct I should add that I'm able to evaluate flags before I call the blocky `identify` call.
- It may be possible that my main app somehow, somewhere blocks the main thread when waiting for LD call to complete, and that's why I'm able to reproduce the issue by calling blockingAwait() and I'm not by calling subscribe and doing it asynchronously. I'll keep investigating that.

**Comment by gwhelanLD** [comment]
Hi @mateuszkwiecinski,

Thanks for the detailed report. I'll be taking a look to try and replicate this issue. 

I have one question about the code sample provided, but it may be uninformed due to lack of experience with rxJava. I see that the sample is catching `InterruptedException` and in the catch block setting an error on the emitter. If `future.get()` were to instead throw a `ExecutionException` would an `await` on the resulting `Completable` complete? 

If there's any replication code you cannot share publicly, but could share privately, you could reach out to our support team using this form to shared that privately: 
https://support.launchdarkly.com/hc/en-us/requests/new
This could also allow us to test against your particular project if any behavior is dependent on the flag payload. 

Thanks,
@gwhelanLD

**Comment by mateuszkwiecinski** [comment]
Hi @gwhelanLD!
Here's the sample project: https://github.com/mateuszkwiecinski/launchdarkly_bug 

There are 2 interesting places: 
1. [Line where the app hangs and never completes](https://github.com/mateuszkwiecinski/launchdarkly_bug/blob/master/app/src/main/java/pls/help/launchdarkly/MainActivity.kt#L36)
2. [build.gradle](https://github.com/mateuszkwiecinski/launchdarkly_bug/blob/master/app/build.gradle#L32) where you can fix wrong behavior by simply downgrading to one of previous versions. 

And yeah, I know [that's](https://github.com/mateuszkwiecinski/launchdarkly_bug/blob/master/app/src/main/java/pls/help/launchdarkly/MainActivity.kt#L27) not the most efficient way to initialize this library, but I assume that waiting for initialized client shouldn't break any `identify` logic.

>  I see that the sample is catching InterruptedException and in the catch block setting an error on the emitter

That's strictly [related to RxJava behavior](https://github.com/ReactiveX/RxJava/wiki/What's-different-in-2.0#error-handling) when canceling a downstream. 

> If future.get() were to instead throw a ExecutionException would an await on the resulting Completable complete?

No, it won't complete. The error will be caught by RxJava's exception handler and re-emited using RxJava's `onError` event. In my example you can observe same behavior without using RxJava.

**Comment by gwhelanLD** [comment]
I appreciate the explanation of the rxJava behavior in this case. 

Thanks for providing a replication project, I've been able to replicate the issue. I'll investigate the root cause and update when I know more.

I do note that you should probably avoid publicly disclosing mobile keys when possible.

**Comment by gwhelanLD** [comment]
Inspection of this issue has lead to the `Throttler` class (and interactions with the `LDAwaitFuture` implementation) as the root cause. This was setup to run it's task in the main `Looper` as a workaround for a part of the initialization requiring being run in the main thread preventing calling `init` in a background thread. In recent changes this shouldn't be required anymore, so the change we'll make is to change that to run it's task in a background thread.  

We'll release this change in a new patch release, though it might be delayed slightly as we're also looking to get in changes required for #76 if possible.

**Comment by gwhelanLD** [comment]
We've just released 2.8.4, which addresses the issue for the replication repo you provided. Please give it a try to confirm that the solution works correctly for your full application.

Thanks for your very helpful report and patience for this issue. Don't hesitate to reach out here on the Github issue tracker or through support channels if you run into any more issues. 

Thanks,
@gwhelanLD

**Comment by gwhelanLD** [comment]
Hi @mateuszkwiecinski,

I'm going to close out this issue as the 2.8.4 release fixes the replication repo you provided. 

Thanks,
@gwhelanLD
